File: bench/py/dot_f64.py
import time
n = 1_000_000
seed = 24680
rng = seed
a = []; b = []
for _ in range(n):
    rng = (rng * 1103515245 + 12345) % 2147483648
    a.append(((rng >> 16) % 1000) / 100.0)
    rng = (rng * 1103515245 + 12345) % 2147483648
    b.append(((rng >> 16) % 1000) / 100.0)
t0 = time.perf_counter()
s = 0.0
for i in range(n):
    s += a[i] * b[i]
t1 = time.perf_counter()
print(f"verify {s:.6f}")
print(f"ms {int((t1-t0)*1000)}")

File: bench/cpp/dot_f64.cpp
// bench/cpp/dot_f64.cpp — double 点积（浮点归约），与 MYP 同算法
#include <chrono>
#include <cstdio>
#include <vector>

static double dotF64(int n, int seed) {
    std::vector<double> a(n), b(n);
    long long rng = seed;
    for (int i = 0; i < n; ++i) {
        rng = (rng * 1103515245LL + 12345LL) % 2147483648LL;
        a[i] = (double)((rng >> 16) % 1000) / 100.0;
        rng = (rng * 1103515245LL + 12345LL) % 2147483648LL;
        b[i] = (double)((rng >> 16) % 1000) / 100.0;
    }
    double sum = 0.0;
    for (int i = 0; i < n; ++i) sum += a[i] * b[i];
    return sum;
}

int main() {
    const int n = 1000000;
    auto t0 = std::chrono::steady_clock::now();
    double v = dotF64(n, 24680);
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("verify %.6f\nms %.0f\n", v, ms);
    return 0;
}
